README.md
# Async Crypter

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/miladiir/node-async-crypter-ts/Node.js%20CI)](https://github.com/Miladiir/node-async-crypter-ts/actions/workflows/node.js.yml?query=branch%3Amain) [![License](https://img.shields.io/npm/l/async-crypter)](https://github.com/Miladiir/node-async-crypter-ts/blob/main/LICENSE) [![NPM](https://img.shields.io/npm/v/async-crypter)](https://www.npmjs.com/package/async-crypter) [![Sponsors](https://img.shields.io/github/sponsors/Miladiir)](https://github.com/sponsors/Miladiir)

Easy, safe and powerful asynchronous AES-256-GCM encryption for NodeJS.

### Table of contents

- [Motivation](#motivation)
- [Usage](#usage)
- [Installation](#installation)
- [Updates](#updates)

### License & Support

This package is licensed under the Unlicense. Please consider supporting me and my projects by
becoming [a sponsor](https://github.com/sponsors/Miladiir).

## Motivation

I was trying to find out how to encrypt files and text using NodeJS. I managed to find my way through the API, but
implementing this was not trivial. I therefore decided to open source what I found out.

Other packages sometimes had implementation errors that can be actually dangerous in production (iv reuse), where
missing features (no aad), or lacked in code quality. One common thing I found while looking at other packages was, that
Buffers where stringified before encryption, which did not make much sense to me.

In the vast sea of packages, there is probably something just as good as this. But this is mine and I am happy with the
way it turned out. I hope you can enjoy and put it to good use as well.

## Usage

To encrypt some text, an image, some file or anything else you can store in a Buffer, first instantiate a new instance
of `Crypter`. At this point you should have a pretty strong password, that will be used to encrypt your data.

```typescript
import {Crypter} from "async-crypter";
// or import xyz from "async-crypter" if you want to use the default export.

const crypterOne = new Crypter("Super Secret Password");

const bufferedSecret = Buffer.from("Another Secret");
const crypterTwo = new Crypter(bufferedSecret);
```

Two instances of `Crypter` that were instantiated with the same secret will behave in the same way. You can now use your
instance to encrypt and decrypt buffers. Just make sure that you do not modify the encrypted buffer, otherwise you will
probably destroy the data stored in it. If you need to modify your data in a meaningful way, first decrypt and then
encrypt again. The encrypted buffer should only be stored or transmitted, but not modified.

Encryption and decryption use the secret, that is stored in the crypter instance.

```typescript
// Pseudocode
async () => {
    const someBuffer = Buffer.from(...);
    const encryptedData = await crypter.encrypt(someBuffer);
...
    const decryptedData = await crypter.decrypt(encryptedData);
}
```

You can also pass additional authenticated data (AAD) as the optional second parameter in both `encrypt` and `decrypt`.
AAD can be used as a second factor. If you loose either the secret or the AAD, the Buffer cannot be decrypted. AAD are
not that useful if they stay the same for all encryption operations and their usage depends on the use-case. I therefore
left them out of the class and you will have to pass them correctly. AAD are just another Buffer with information that
you perceive as a second important authentication factor for the purpose of encryption. AAD might be the users ID, or
the original path of the file, or a combination of multiple criteria. This is up to you to implement.

**Remember**: You will need the same secret, the full and unmodified encrypted Buffer and the same AAD to decrypt your
Buffer again.

## Properties

Buffers that are encrypted using this library can be decrypted using standard cryptograhpy libraries, e.g. openssl, and
the information about the encrypted Buffer layout.

- Data is encrypted using nodejs crypto apis linked to openssl and the aes-256-gcm cipher.
- The output buffer contains the salt (64 bytes), initalization vector (16 bytes), authentication tag (16 bytes) and of
  course the encrypted data (remaining bytes).

The actual encryption key is generated using `pbkdf2` with the original secret (stored in `Crypter` instance), a
randomly generated salt (the one that is later prepended to the encrypted Buffer), an iteration count of 100.000, a
keylength of 32 and `sha512` as the digest. In addition to this ever changing encryption key, additional safety is
provided by a changing initialization vector per encryption. Combined this ensures that every encryption is completely
non-predictable. Even if you use the same password and the same Buffer, you will generate distinct resulting encrypted
Buffers. As explained in [Usage](#usage), additional authenticated data can be used to further influence the encryption.
Unlike iv, salt and tag, the add are not included in the encrypted buffer. Think of aad more like a second password.

I unfortunately cannot find much information on how the authentication tag is formed. I must warn you that the
authentication tag is probably implementation specific to NodeJS and that you can probably only form or verify one using
the crypto APIs of node.

## Installation

`npm i --save async-crypter`

The [npm package](https://www.npmjs.com/package/async-crypter) contains the transpiled JavaScript code and TypeScript
typings. Both are generated from the TypeScript source code. You do not need TypeScript to use this package.

## Updates

**async-crypter** follows [Semantic Versioning 2.0.0](https://semver.org/#semantic-versioning-200). This means that you
can decide based on the version number of the package if manual update intervention is required. Head over
to [Github Releases](https://github.com/Miladiir/node-async-crypter-ts/releases) or check the CHANGELOG file for changes
between versions. In most cases *npm* will take care of updates for you automatically with `npm update` or an
alternative of your choice.

(Not recommended)
`npm i --save async-crypter@latest` will force the package to the latest version in any case.


CHANGELOG
Tag v1.0.0:
Features:
- Added Crypter, an asynchronous AES-256-GCM Buffer encryption and decryption helper class
Documentation:
- Crypter


test/index.spec.ts
import {Crypter as CrypterA} from "../src";
import CrypterB from "../src";
import {expect} from "chai";

describe("index", (): void => {

    it("should be the same imported module", (): void => {

        expect(CrypterA).to.equal(CrypterB);
    });
});

test/Crypter.spec.ts
import {expect} from "chai";
import {Crypter} from "../src/Crypter";
import {DecryptionError, EncryptionError} from "../src";

describe("Crypter", (): void => {

    const secret = "This is a very secure secret and I will get mad if you think otherwise.";
    const value = "Try to steal me. I dare you.";

    describe("constructor", (): void => {

        it("should accept non empty string", (): void => {

            expect(new Crypter(secret)).to.be.instanceOf(Crypter);
        });

        it("should not accept empty string", (): void => {

            expect(() => new Crypter("")).to.throw(Error);
        });

        it("should accept non empty Buffer", (): void => {

            expect(new Crypter(Buffer.from(secret))).to.be.instanceOf(Crypter);
        });

        it("should not accept empty Buffer", (): void => {

            expect(() => new Crypter(Buffer.alloc(0))).to.throw(Error);
        });

        it("should not accept anything else", (): void => {

            // @ts-expect-error Explicitly test incompatible types
            expect(() => new Crypter(null)).to.throw(Error);
            // @ts-expect-error Explicitly test incompatible types
            expect(() => new Crypter(undefined)).to.throw(Error);
            // @ts-expect-error Explicitly test incompatible types
            expect(() => new Crypter()).to.throw(Error);
            // @ts-expect-error Explicitly test incompatible types
            expect(() => new Crypter(0)).to.throw(Error);
        });
    });

    describe("encrypt", (): void => {

        it("should generate distinct output files (distinct instances)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter1 = new Crypter(secret);
            const crypter2 = new Crypter(secret);
            const encrypted1 = await crypter1.encrypt(buffer);
            const encrypted2 = await crypter2.encrypt(buffer);
            expect(encrypted1.toString()).to.not.equal(encrypted2.toString());
        });

        it("should generate distinct output files (same instance)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted1 = await crypter.encrypt(buffer);
            const encrypted2 = await crypter.encrypt(buffer);
            expect(encrypted1.toString()).to.not.equal(encrypted2.toString());
        });

        it("should work for an empty Buffer", async (): Promise<void> => {

            const buffer = Buffer.alloc(0);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            expect(encrypted.byteLength).to.be.at.least(64 + 16 + 16);
        });

        it("should fail for non Buffer values", async (): Promise<void> => {

            const crypter = new Crypter(secret);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt(123)).to.be.instanceOf(EncryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt("")).to.be.instanceOf(EncryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt(null)).to.be.instanceOf(EncryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt(undefined)).to.be.instanceOf(EncryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt()).to.be.instanceOf(EncryptionError);
        });

        it("should fail for empty Buffer aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const aad = Buffer.alloc(0);
            expect(await crypter.encrypt(buffer, aad)).to.be.instanceOf(EncryptionError);
        });

        it("should fail for any non Buffer type of aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt(buffer, 0)).to.be.instanceOf(EncryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.encrypt(buffer, new Date())).to.be.instanceOf(EncryptionError);
        });
    });

    describe("decrypt", (): void => {

        it("should be able to decrypt data (same class instance)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            const decrypted = await crypter.decrypt(encrypted);
            expect(buffer.toString()).to.equal(decrypted.toString());
        });

        it("should be able to decrypt data (same password)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter1 = new Crypter(secret);
            const encrypted = await crypter1.encrypt(buffer) as Buffer;
            const crypter2 = new Crypter(secret);
            const decrypted = await crypter2.decrypt(encrypted);
            expect(buffer.toString()).to.equal(decrypted.toString());
        });

        it("should fail for non Buffer values", async (): Promise<void> => {

            const crypter = new Crypter(secret);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt(123)).to.be.instanceOf(DecryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt("")).to.be.instanceOf(DecryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt(null)).to.be.instanceOf(DecryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt(undefined)).to.be.instanceOf(DecryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt()).to.be.instanceOf(DecryptionError);
        });

        it("should be able to decrypt data (same aad)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const additionalData = Buffer.from("Some additional data, like a userId, a random secret or something");
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer, additionalData) as Buffer;
            const decrypted = await crypter.decrypt(encrypted, additionalData);
            expect(buffer.toString()).to.equal(decrypted.toString());
        });

        it("should not be able to decrypt data with wrong password", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter1 = new Crypter(secret);
            const encrypted = await crypter1.encrypt(buffer) as Buffer;
            const wrongSecret = secret.replace("secure", "insecure");
            const crypter2 = new Crypter(wrongSecret);
            const decrypted = await crypter2.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with wrong aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const additionalData = Buffer.from("Some additional data, like a userId, a random secret or something");
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer, additionalData) as Buffer;
            const decrypted = await crypter.decrypt(encrypted, buffer);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with missing aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const additionalData = Buffer.from("Some additional data, like a userId, a random secret or something");
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer, additionalData) as Buffer;
            const decrypted = await crypter.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with aad when not encrypted using aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const additionalData = Buffer.from("Some additional data, like a userId, a random secret or something");
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            const decrypted = await crypter.decrypt(encrypted, additionalData);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with corrupted data (salt)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            encrypted.writeUInt8(0x00, 12);
            const decrypted = await crypter.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with corrupted data (iv)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            encrypted.writeUInt8(0x00, 64 + 3);
            const decrypted = await crypter.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with corrupted data (tag)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            encrypted.writeUInt8(0x00, 64 + 16 + 2);
            const decrypted = await crypter.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should not be able to decrypt data with corrupted data (data)", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            encrypted.writeUInt8(0x00, 64 + 16 + 16 + 12);
            const decrypted = await crypter.decrypt(encrypted);
            expect(decrypted).to.be.instanceOf(DecryptionError);
        });

        it("should fail for empty Buffer aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const aad = Buffer.alloc(0);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            expect(await crypter.decrypt(encrypted, aad)).to.be.instanceOf(DecryptionError);
        });

        it("should fail for any non Buffer type of aad", async (): Promise<void> => {

            const buffer = Buffer.from(value);
            const crypter = new Crypter(secret);
            const encrypted = await crypter.encrypt(buffer) as Buffer;
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt(encrypted, 0)).to.be.instanceOf(DecryptionError);
            // @ts-expect-error Explicitly test incompatible types
            expect(await crypter.decrypt(encrypted, new Date())).to.be.instanceOf(DecryptionError);
        })
    });
});

src/Crypter.ts
import {createCipheriv, createDecipheriv, pbkdf2, randomBytes} from "crypto";
import {promisify} from "util";
import {EncryptionError} from "./errors/EncryptionError";
import {ErrorMessage} from "./errors/ErrorMessage";
import {DecryptionError} from "./errors/DecryptionError";

/**
 * Performs AES-256-GCM encryption and decryption of buffers.
 * The encrypted buffers contain the salt (64 bytes), the iv (16 bytes), the tag (16 bytes)
 * and finally the encrypted value (rest of the buffer).
 */
export class Crypter {

    /**
     * The algorithm in use.
     */
    private static readonly algorithm = 'aes-256-gcm';

    /**
     * The length of an initialization vector in bytes.
     */
    private static readonly ivLength = 16;

    /**
     * The length of a salt in bytes.
     */
    private static readonly saltLength = 64;

    /**
     * The length of an authentication tag in bytes.
     */
    private static readonly tagLength = 16;

    /**
     * The offset of the authentication tag in bytes.
     */
    private static readonly tagPosition = Crypter.saltLength + Crypter.ivLength;

    /**
     * The offset of the encrypted value in bytes.
     */
    private static readonly encryptedPosition = Crypter.tagPosition + Crypter.tagLength;

    /**
     * Promisified crypto.pbkdf2
     */
    private static readonly pbkdf2 = promisify(pbkdf2);

    /**
     * The secret used to encrypt and decrypt values by this instance.
     */
    readonly #secret: Buffer;

    /**
     * Initiate a new Crypter instance with the specified secret.
     * @param secret The secret used to encrypt and decrypt values.
     * @throws Error when passed anything but a non-empty string or Buffer.
     */
    public constructor(secret: string | Buffer) {

        if (secret instanceof Buffer && secret.length !== 0) {
            this.#secret = secret;
        } else if (typeof secret === "string" && secret.length !== 0) {
            this.#secret = Buffer.from(secret);
        } else {
            throw new Error(ErrorMessage.INVALID_SECRET);
        }
    }

    /**
     * Encrypt a buffer using AES256-GCM with a new generated salt and a random
     * initialization vector.
     * @param value The value to encrypt.
     * @param additionalAuthenticatedData Additional data that influences encryption outcome.
     * @returns The uniquely and symmetrically encrypted value.
     */
    public async encrypt(value: Buffer, additionalAuthenticatedData?: Buffer): Promise<Buffer | EncryptionError> {

        if (!(value instanceof Buffer)) {

            return new EncryptionError(ErrorMessage.INVALID_ENCRYPTION_VALUE);
        }
        const iv = randomBytes(Crypter.ivLength);
        const salt = randomBytes(Crypter.saltLength);
        const key = await this.generateKey(salt);
        const cipher = createCipheriv(Crypter.algorithm, key, iv);
        if (additionalAuthenticatedData != null) {
            if (additionalAuthenticatedData instanceof Buffer && additionalAuthenticatedData.length !== 0) {
                cipher.setAAD(additionalAuthenticatedData);
            } else {

                return new EncryptionError(ErrorMessage.INVALID_AAD);
            }
        }
        let updatedBuffer, finalBuffer, authTag: Buffer;
        try {
            updatedBuffer = cipher.update(value);
            finalBuffer = cipher.final();
            authTag = cipher.getAuthTag();
        } catch (e) {

            return new EncryptionError("Unknown error");
        }

        return Buffer.concat([salt, iv, authTag, updatedBuffer, finalBuffer]);
    }

    /**
     * Decrypts a buffer containing the salt, initialization vector, tag and encrypted value.
     * @param value An encrypted value buffer.
     * @param additionalAuthenticatedData Additional data that was used to influence encryption outcome.
     * @returns The decrypted buffer containing the original value.
     */
    public async decrypt(value: Buffer, additionalAuthenticatedData?: Buffer): Promise<Buffer | DecryptionError> {

        if (!(value instanceof Buffer) || value.byteLength < Crypter.encryptedPosition) {

            return new DecryptionError(ErrorMessage.INVALID_DECRYPTION_VALUE);
        }

        const salt = value.slice(0, Crypter.saltLength);
        const key = await this.generateKey(salt);
        const iv = value.slice(Crypter.saltLength, Crypter.tagPosition);
        const tag = value.slice(Crypter.tagPosition, Crypter.encryptedPosition);
        const encrypted = value.slice(Crypter.encryptedPosition);
        const decipher = createDecipheriv(Crypter.algorithm, key, iv);
        if (additionalAuthenticatedData != null) {
            if (additionalAuthenticatedData instanceof Buffer && additionalAuthenticatedData.length !== 0) {
                decipher.setAAD(additionalAuthenticatedData);
            } else {

                return new DecryptionError(ErrorMessage.INVALID_AAD);
            }
        }
        decipher.setAuthTag(tag);
        let updatedBuffer, finalBuffer: Buffer;
        try {
            updatedBuffer = decipher.update(encrypted);
            finalBuffer = decipher.final();
        } catch (e) {

            return new DecryptionError(ErrorMessage.DECRYPTION_FAILED);
        }

        return Buffer.concat([updatedBuffer, finalBuffer]);
    }

    /**
     * Generate an encryption key using the provided salt and the secret of the class.
     * @param salt A salt, either a new random sequence or stored in an encrypted buffer.
     * @returns An encryption key used to encrypt or decrypt a buffer.
     */
    private generateKey(salt: Buffer): Promise<Buffer> {

        return Crypter.pbkdf2(this.#secret, salt, 100000, 32, 'sha512');
    }
}